### DISCUSSION TITLE
dismiss is not working after multiple views were closed by UIKit

### DISCUSSION THREAD

**Discussion by safarafa** [pointfreeco/swift-composable-architecture#2400] [General]
### Description

Child view is shown using `@PresentationState` in parent view, and both of them are shown in hybrid application, using `UIHostingController` shown as modal in new `UINavigationController`, and this whole modal view is dismissed (by UIKit method). When we'll show child view again, TCA's `dismiss` method is not working - actions are sent, the state is set to nil, but the corresponding view is still visible. It might be bug, because apparently the state is not nil after _cancelling_ process.

[Sample app](https://github.com/safarafa/tca-dismiss-issue/tree/main)

### Checklist

- [ ] I have determined whether this bug is also reproducible in a vanilla SwiftUI project.
- [X] If possible, I've reproduced the issue using the `main` branch of this package.
- [X] This issue hasn't been addressed in an [existing GitHub issue](https://github.com/pointfreeco/swift-composable-architecture/issues) or [discussion](https://github.com/pointfreeco/swift-composable-architecture/discussions).

### Expected behavior

Using `.dismiss()` should remove view and set state to nil.

### Actual behavior

Using `.dismiss()` is not removing view and the state is not nil.

### Steps to reproduce

1. Download and run sample app, on device or simulator.
2. Tap "Dismiss test"
3. Tap "Tap to see first level child"
4. Tap "Dismiss using UIKit"
5. Tap "Dismiss test"
6. Tap "Dismiss this Child" 
Reproduction rate: 3/5 (sometime dismiss works).

### The Composable Architecture version information

1.1.0 and below

### Destination operating system

iOS 14

### Xcode version information

14.3.1

### Swift Compiler version information

```shell
Apple Swift version 5.8.1
```

**Comment by mbrandonw** [comment]
Hi @safarafa, this is the expected behavior of `\.dismiss`. All it does is find the closest optional state that is driving some kind of navigation and `nil`s it out in order to undo the navigation. So, in your case it finds `@PresentationState var child` and `nil`s out the child state, causing the screen to be popped off the stack.

It's worth noting that this is exactly how `@Environment(\.dismiss)` works in vanilla SwiftUI.

Since this is not an issue with the library I will convert this to a discussion. Feel free to continue discussing over there if you have more questions.

**Reply by safarafa** [reply]
I have to be sure you understand my issue.
After some interaction I've got something like that:
`UINavigationController -> UIViewController -> UINavigationController (shown modally) -> Parent View -> Child View`
if I tap "Dismiss this Child" I would have:
`UINavigationController -> UIViewController -> UINavigationController (shown modally) -> Parent View`
but I tap "Dismiss using UIKit" and I have 
`UINavigationController -> UIViewController`
and then, after dismissing using UIKit, I interact to once again have: 
`UINavigationController -> UIViewController -> UINavigationController (shown modally) -> Parent View -> Child View`
when I tap "Dismiss this Child" it is not working - it was working before UIKit dismissal of whole modal.

**Reply by stephencelis** [reply]
Hi @safarafa! Brandon and I looked at this earlier and it looks like it might be a SwiftUI bug with observing too much state.

If you remove the `WithViewStore(observe: { $0 })` wrapping the `NavigationStackStore` here...

https://github.com/safarafa/tca-dismiss-issue/blob/c9033210009d852d6f75a36375d68fabfe735377/dismiss-test/Contents.swift#L33-L45

...and invoke `self.store.send(.itemTapped)` instead of `viewStore.send(.itemTapped)` it seems to fix the problem.

**Reply by safarafa** [reply]
Thank you for a fast response. It worked, also in the larger project in which I observed the same behaviour.

**Comment by szymhan** [comment]
Hi all, I also reproduced the unexpected behaviour and recorded short video ⬇️
![dismiss_error](https://github.com/pointfreeco/swift-composable-architecture/assets/14939026/991f05c7-a143-4b76-8ee9-c88cec9d0ec4)

It seems, that calling UIKit's `dismiss(animated:)` is breaking something around PresentationStore, but I'm not able to diagnose, what specifically is happening there 🤔
